File: examples/expected_output.go
// Expected Go output for simple_class.cpp

package main

import "fmt"

type Point struct {
    X int32
    Y int32
}

func NewPoint(x, y int32) Point {
    return Point{X: x, Y: y}
}

func (p *Point) GetX() int32 {
    return p.X
}

func (p *Point) GetY() int32 {
    return p.Y
}

func (p *Point) SetX(newX int32) {
    p.X = newX
}

func (p *Point) SetY(newY int32) {
    p.Y = newY
}

func (p *Point) DistanceSquared() int32 {
    return p.X*p.X + p.Y*p.Y
}

func (p *Point) Translate(dx, dy int32) {
    p.X += dx
    p.Y += dy
}

type Rectangle struct {
    TopLeft     Point
    BottomRight Point
}

func NewRectangle(tl, br Point) Rectangle {
    return Rectangle{
        TopLeft:     tl,
        BottomRight: br,
    }
}

func (r *Rectangle) Width() int32 {
    return r.BottomRight.GetX() - r.TopLeft.GetX()
}

func (r *Rectangle) Height() int32 {
    return r.BottomRight.GetY() - r.TopLeft.GetY()
}

func (r *Rectangle) Area() int32 {
    return r.Width() * r.Height()
}

type Resource struct {
    Data []int32
    Size uint
}

func NewResource(n uint) *Resource {
    return &Resource{
        Data: make([]int32, n),
        Size: n,
    }
}

func (r *Resource) Get(index uint) int32 {
    return r.Data[index]
}

func (r *Resource) Set(index uint, value int32) {
    r.Data[index] = value
}

func (r *Resource) GetSize() uint {
    return r.Size
}

type SharedData struct {
    Message string
}

func NewSharedData(msg string) *SharedData {
    return &SharedData{
        Message: msg,
    }
}

func (s *SharedData) GetMessage() string {
    return s.Message
}

func (s *SharedData) SetMessage(msg string) {
    s.Message = msg
}

func main() {
    p1 := NewPoint(10, 20)
    p2 := NewPoint(30, 40)

    rect := NewRectangle(p1, p2)
    a := rect.Area()

    res := NewResource(100)
    res.Set(0, 42)

    data := NewSharedData("Hello, World!")

    fmt.Printf("Area: %d\n", a)
    fmt.Printf("Resource[0]: %d\n", res.Get(0))
    fmt.Printf("Message: %s\n", data.GetMessage())
}

File: examples/expected_output.rs
// Expected Rust output for simple_class.cpp

pub struct Point {
    x: i32,
    y: i32,
}

impl Point {
    pub fn new(x: i32, y: i32) -> Self {
        Self { x, y }
    }

    pub fn get_x(&self) -> i32 {
        self.x
    }

    pub fn get_y(&self) -> i32 {
        self.y
    }

    pub fn set_x(&mut self, new_x: i32) {
        self.x = new_x;
    }

    pub fn set_y(&mut self, new_y: i32) {
        self.y = new_y;
    }

    pub fn distance_squared(&self) -> i32 {
        self.x * self.x + self.y * self.y
    }

    pub fn translate(&mut self, dx: i32, dy: i32) {
        self.x += dx;
        self.y += dy;
    }
}

pub struct Rectangle {
    top_left: Point,
    bottom_right: Point,
}

impl Rectangle {
    pub fn new(tl: Point, br: Point) -> Self {
        Self {
            top_left: tl,
            bottom_right: br,
        }
    }

    pub fn width(&self) -> i32 {
        self.bottom_right.get_x() - self.top_left.get_x()
    }

    pub fn height(&self) -> i32 {
        self.bottom_right.get_y() - self.top_left.get_y()
    }

    pub fn area(&self) -> i32 {
        self.width() * self.height()
    }
}

pub struct Resource {
    data: Vec<i32>,
    size: usize,
}

impl Resource {
    pub fn new(n: usize) -> Self {
        Self {
            data: vec![0; n],
            size: n,
        }
    }

    pub fn get(&self, index: usize) -> i32 {
        self.data[index]
    }

    pub fn get_mut(&mut self, index: usize) -> &mut i32 {
        &mut self.data[index]
    }

    pub fn get_size(&self) -> usize {
        self.size
    }
}

use std::rc::Rc;
use std::cell::RefCell;

pub struct SharedData {
    message: Rc<RefCell<String>>,
}

impl SharedData {
    pub fn new(msg: &str) -> Self {
        Self {
            message: Rc::new(RefCell::new(msg.to_string())),
        }
    }

    pub fn get_message(&self) -> String {
        self.message.borrow().clone()
    }

    pub fn set_message(&mut self, msg: &str) {
        *self.message.borrow_mut() = msg.to_string();
    }
}

fn main() {
    let p1 = Point::new(10, 20);
    let p2 = Point::new(30, 40);

    let rect = Rectangle::new(p1, p2);
    let a = rect.area();

    let mut res = Resource::new(100);
    *res.get_mut(0) = 42;

    let mut data = SharedData::new("Hello, World!");

    println!("Area: {}", a);
    println!("Resource[0]: {}", res.get(0));
    println!("Message: {}", data.get_message());
}
